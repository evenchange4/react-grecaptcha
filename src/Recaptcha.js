import React, { PropTypes } from 'react';
import c from 'classnames';
import omit from 'lodash.omit';

// Constants
const ID = '_grecaptcha.element.id';
const CALLBACK_NAME = '_grecaptcha.data-callback';
const EXPIRED_CALLBACK_NAME = '_grecaptcha.data-expired-callback';

const removeChild = elem => elem.parentNode && elem.parentNode.removeChild(elem);

class Recaptcha extends React.Component {
  static propTypes = {
    // Required
    sitekey: PropTypes.string.isRequired,
    callback: PropTypes.func.isRequired,
    expiredCallback: PropTypes.func.isRequired,

    // Options
    className: PropTypes.string,
    invisible: PropTypes.bool,
    locale: PropTypes.string,
  }

  static defaultProps = {
    locale: 'en',
    className: undefined,
    invisible: false,
  }

  componentDidMount() {
    const { locale, callback, expiredCallback } = this.props;

    // 1. Async lazy load
    const head = document.head || document.getElementsByTagName('head')[0];
    const script = document.createElement('script');
    script.id = ID;
    script.src = `https://www.google.com/recaptcha/api.js?hl=${locale}`;
    script.type = 'text/javascript';
    script.async = true;
    script.defer = true;
    script.onerror = (oError) => {
      throw new URIError(`The script ${oError.target.src} is not accessible.`);
    };
    head.appendChild(script);

    // 2. Expose callback function to window object
    window[CALLBACK_NAME] = callback;
    window[EXPIRED_CALLBACK_NAME] = expiredCallback;
  }

  componentWillUnmount() {
    removeChild(document.getElementById(ID));
  }

  render() {
    const { className, sitekey, invisible, ...otherProps } = this.props;
    const props = {
      ...omit(otherProps, ['callback', 'expiredCallback', 'locale']),
      className: c('g-recaptcha', className),
      'data-sitekey': sitekey,
      'data-callback': CALLBACK_NAME,
      'data-expired-callback': EXPIRED_CALLBACK_NAME,
      ...(invisible && { 'data-size': 'invisible' }),
    };

    return (
      <div {...props} />
    );
  }
}

export default Recaptcha;
