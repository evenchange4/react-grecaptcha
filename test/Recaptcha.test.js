import React from 'react';
import renderer from 'react-test-renderer';
import Recaptcha from '../src/Recaptcha';

describe('<Recaptcha />', () => {
  const wrapper = renderer.create(
    <Recaptcha
      sitekey="6LedJiATAAAAAJvpLFxW9dSiEGu8rGyoOIuTsvtg"
      callback={() => {
      }}
      expiredCallback={() => {
      }}
    />,
  );

  const invisibleWrapper = renderer.create(
    <Recaptcha
      sitekey="6LedJiATAAAAAJvpLFxW9dSiEGu8rGyoOIuTsvtg"
      callback={() => {
      }}
      expiredCallback={() => {
      }}
      invisible
    />,
  );

  it('should render with `g-recaptcha` className', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render without non-HTML component properties', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render invisible version', () => {
    expect(invisibleWrapper.toJSON()).toMatchSnapshot();
  });
});
