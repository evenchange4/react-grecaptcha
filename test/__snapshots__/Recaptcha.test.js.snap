// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recaptcha /> should render invisible version 1`] = `
<div
  className="g-recaptcha"
  data-callback="_grecaptcha.data-callback"
  data-expired-callback="_grecaptcha.data-expired-callback"
  data-sitekey="6LedJiATAAAAAJvpLFxW9dSiEGu8rGyoOIuTsvtg"
  data-size="invisible"
/>
`;

exports[`<Recaptcha /> should render with \`g-recaptcha\` className 1`] = `
<div
  className="g-recaptcha"
  data-callback="_grecaptcha.data-callback"
  data-expired-callback="_grecaptcha.data-expired-callback"
  data-sitekey="6LedJiATAAAAAJvpLFxW9dSiEGu8rGyoOIuTsvtg"
/>
`;

exports[`<Recaptcha /> should render without non-HTML component properties 1`] = `
<div
  className="g-recaptcha"
  data-callback="_grecaptcha.data-callback"
  data-expired-callback="_grecaptcha.data-expired-callback"
  data-sitekey="6LedJiATAAAAAJvpLFxW9dSiEGu8rGyoOIuTsvtg"
/>
`;
